### PR TITLE
Dynamic retreat node specification

### PR DESCRIPTION
**Pull request by sakura-quest**
Adds support for user specified custom retreat nodes, to be able to vary the number of battles during a sortie based on the route taken as described in Issue #156.
After some time figuring out the js dependencies for the browser UI config generator page, I was able to transpile the react code. However, I could only get it to do so in developer mode with the command 
`webpack --env.dev --mode=development`. It seems like the version of webpack I am using is too new since I get an error
`Error: webpack.optimize.UglifyJsPlugin has been removed, please use config.optimization.minimize instead.`  I am not a js developer, so it is very possible that I am doing something wrong, in which case I would appreciate any pointers on how I should be transpiling. 

Hence, this pull request does not include changes to the transpiled bundle.js. However, if you can let me know what version of webpack I should be using, as well as the webpack CLI flags, I'll gladly transpile and include the changes to bundle.js in my PR.

### CODE REVIEW

**Comment by sakura-quest** [comment]
Sample config.ini with this change that I've tried for testing.
```
[General]
Program: Chrome
JSTOffset: -16
Pause: False

[ScheduledSleep]
SleepEnabled: True
SleepStartTime: 0115
SleepLength: 10
ExpeditionSleepEnabled: False
ExpeditionSleepStartTime: 0000
ExpeditionSleepLength: 3
CombatSleepEnabled: False
CombatSleepStartTime: 0000
CombatSleepLength: 5

[Expeditions]
Enabled: True
Fleet2: 2
Fleet3: 5
Fleet4: 21

[PvP]
Enabled: False

[Combat]
Enabled: True
Engine: live
Map: 4-3
FleetMode: 
CombatNodes: 3
NodeSelects: 
Formations: D:line_abreast,F:line_abreast,J:line_ahead,K:line_abreast
NightBattles: 1:False,2:False,3:False
RetreatLimit: heavy
RepairLimit: heavy
RepairTimeLimit: 0029
LBASGroups: 
LBASGroup1Nodes: 
LBASGroup2Nodes: 
LBASGroup3Nodes: 
ForceRetreatNodes: D,K
MiscOptions: CheckFatigue

[ShipSwitcher]
Enabled: False
Slot1Criteria: 
Slot1Ships: 
Slot2Criteria: 
Slot2Ships: 
Slot3Criteria: 
Slot3Ships: 
Slot4Criteria: 
Slot4Ships: 
Slot5Criteria: 
Slot5Ships: 
Slot6Criteria: 
Slot6Ships: 

[Quests]
Enabled: False
```

**Comment by sakura-quest** [comment]
Same as #157 but used the wrong email.

File: docs-src/src/components/BodyConfigCombat.jsx [hunk 1]
```diff
@@ -173,6 +173,23 @@ class BodyConfigCombat extends PureComponent {
     this.setState({ combatLBASGroups: value })
   }
 
+  handleCombatForceRetreatNodeAdd = (node) => {
```
Comment (mrmin123) [inline docs-src/src/components/BodyConfigCombat.jsx:176]: This is a lot of code for what could basically be replaced with the Multi Select component (see the Expeditions dropdown multiselects).

File: docs-src/src/components/BodyConfigCombat.jsx [hunk 2]
```diff
@@ -662,6 +684,51 @@ class BodyConfigCombat extends PureComponent {
...
+                simpleValue={true}
+                name='combatForceRetreatNode1'
+                value={combatForceRetreatNode1}
+                options={NODES}
+                onChange={value => this.setState({ combatForceRetreatNode1: value })}
+                disabled={!combatEnabled}
+                fullWidth />
+            </FormControl>
+          </Grid>
+          <Grid item xs={4} sm={1} className={classes.formGridButton}>
```
Comment (mrmin123) [inline docs-src/src/components/BodyConfigCombat.jsx:703]: Again, the button and collection Multi select can all be replaced by making the previous Select field the Multi field (and the source of truth).

File: kcauto-kai.sikuli/combat.sikuli/mapData.py [hunk 3]
```diff
@@ -139,10 +139,11 @@ def resolve_continue_sortie(self, node):
             node (Node): Node object for the node the fleet is at
 
         Returns:
-            bool: True if sortie shoud be continued, False otherwise
+            bool: True if sortie should be continued, False otherwise
         """
         if node:
-            if 'retreat' in node.types:
+            if node.name in self.config.combat['force_retreat_nodes'] or \
```
Comment (mrmin123) [inline kcauto-kai.sikuli/combat.sikuli/mapData.py:145]: Please format multi-line conditionals in python in the following format:
```python
if (node.name in self.config.combat['force_retreat_nodes']
    or 'retreat' in node.types):
```

**Review by mrmin123 — CHANGES_REQUESTED** [review]
Thanks for the PR!

First things first: please re-stage this PR for the `staging` branch

Second, and on a related note: the webpack version this is built for is 3.8.1 as defined by `package.json`; it should have installed with `yarn install` but webpack is weird so who knows... Either way, the `staging` branch is where all version changes and webUI changes are staged and updated together, so you shouldn't be submitting an updated `bundle.js` with this PR anyway.

Third: I've requested some component changes and stylistic changes to the webUI and python code as part of the PR review, but please refer (and/or add) to the discussion in #156 before making further changes.

**Comment by mrmin123** [comment]
Closing this PR in favor of PR #186, which adds some additional changes that was discussed here and in #156 on top of the changes in this PR.

**Pull request closed without merging**